File: src/lib/filter.cpp


/*
    10/11/2014

    Create a unique filter file (FILTER_FILE) with
    all the single files combined in the list FILTER_LIST

    Derive useful quantities as the AB corrections
*/

#include <getopt.h>  // get option line commands getopt_long
#include <stdlib.h>  // absolute value, exit

#include <cstring>   // Use string in the old c format
#include <fstream>   // print output file
#include <iostream>  // print standard file
#include <sstream>
#include <string>  // use string
#include <vector>  // manipulate vector

// Le Phare
#include "SED.h"
#include "flt.h"      // define the filters
#include "globals.h"  // global variables
#include "keyword.h"  // to read the keyword in the command line/para file

using namespace std;

// START
int main(int argc, char *argv[]) {
  // Declare variables
  vector<flt> vecFlt;

  // to estimate the times needed to run the code
  clock_t t1, t2;
  t1 = clock();

  /*
  ENVIRONMENT VARIABLES LEPHAREDIR and LEPHAREWORK
  */
  get_lephare_env();

  /*
  ANALYSE KEYWORDS
  */

  // List of the keywords to be found in the config file/command line
  string list_keywords[] = {"FILTER_REP", "FILTER_LIST", "TRANS_TYPE",
                            "FILTER_CALIB", "FILTER_FILE"};
  // Number of keywords
  int nb_ref_key = (int)(sizeof(list_keywords) / sizeof(list_keywords[0]));

  // Find the reference keywords within the command line + config file -> store
  // them into a vector
  keymap key_analysed = analyse_keywords(argc, argv, list_keywords, nb_ref_key);

  // For each keyword from the reference list, fill the corresponding variable

  // Configuration file
  string config = key_analysed["c"].value;
  // Repository in which the filters are stored
  string fltRep =
      ((key_analysed["FILTER_REP"]).split_string(lepharedir + "/filt/", 1))[0];
  // filter list with the name of the filter file to be oppened
  vector<string> fltFiles =
      (key_analysed["FILTER_LIST"]).split_string("flt.pb", -99);
  // Use the number of filters as the reference number of keyword (the following
  // list of values should match 1 or this number)
  int ref_size = fltFiles.size();
  // Transmission in energy or photons
  vector<int> transtyp = (key_analysed["TRANS_TYPE"]).split<int>("0", ref_size);
  // calibration depending on the instrument
  vector<int> calibtyp =
      (key_analysed["FILTER_CALIB"]).split<int>("0", ref_size);
  // Output file
  string outputName =
      (key_analysed["FILTER_FILE"]).split<string>("filters", 1)[0];
  string filtfile = lepharework + "/filt/" + outputName + ".dat";
  string filtdoc = lepharework + "/filt/" + outputName + ".doc";

  /*
  INFO PARAMETERS ON SCREEN
  */

  cout << "#######################################" << endl;
  cout << "# Build the filter file with the following options: " << endl;
  cout << "# Config file: " << config << endl;
  cout << "# FILTER_REP: " << fltRep << endl;
  cout << "# FILTER_LIST: ";
  for (int k = 0; k < ref_size; k++) {
    cout << fltFiles[k] << " ";
  }
  cout << endl << "# TRANS_TYPE: ";
  for (int k = 0; k < ref_size; k++) {
    cout << transtyp[k] << " ";
  }
  cout << endl << "# FILTER_CALIB: ";
  for (int k = 0; k < ref_size; k++) {
    cout << calibtyp[k] << " ";
  }
  cout << endl << "# FILTER_FILE: " << filtfile << endl;
  cout << "# FILTER_FILE.doc: " << filtdoc << endl;
  cout << "#######################################" << endl;

  /*
  READ THE FILTER DATA FILE
  */

  // Loop over each filter from the FILTER_LIST
  for (int k = 0; k < ref_size; k++) {
    // Open this filter file
    string fltFile = fltRep + "/" + fltFiles[k];
    // Generate one object "flt"
    flt oneFilt(k, fltFile, transtyp[k], calibtyp[k]);
    // store all filters in a vector
    vecFlt.push_back(oneFilt);
  }

  /*
  WRITE THE FILTER FILE
  */

  // function to write the filters
  write_output_filter(filtfile, filtdoc, vecFlt);

  /*
  DONE
  */

  // display the time needed to run the code
  t2 = clock();
  float diff(((float)t2 - (float)t1) / CLOCKS_PER_SEC);
  cout << "Times to run the code in sec: " << diff << endl;

  return 0;
}

File: src/lephare/filter.py
import os
from contextlib import suppress

from ._lephare import flt, write_output_filter
from .runner import Runner

__all__ = [
    "Filter",
]

config_keys = {
    "verbose": "increase onscreen verbosity",
    "FILTER_REP": "path to repository where filter files are searched",
    "FILTER_LIST": "list of filter files, to be searched in FILTER_REP if relative",
    "TRANS_TYPE": "Transmission curve type: 0[def] for energy, 1 for photon nb",
    "FILTER_CALIB": "calibration system for the filter:\
    0[def]: fnu=cst 1: nu.fnu=cst 2: fnu=nu 3: fnu=Black Body @ T=10000K\
    4: for MIPS (leff with nu fnu=ctt and flux with BB @ 10000K",
    "FILTER_FILE": "output filter filename, will be saved in $LEPHAREWORK/filt/",
}


class Filter(Runner):
    """
    The specific arguments to the Filter class are

    verbose
                increase onscreen verbosity
    FILTER_REP
                path to repository where filter files are searched
    FILTER_LIST
                list of filter files, to be searched in FILTER_REP if relative
    TRANS_TYPE
                Transmission curve type: 0[def] for energy, 1 for photon nb
    FILTER_CALIB
                calibration system for the filter:
                  0[def]: fnu=cst
                  1: nu.fnu=cst
                  2: fnu=nu
                  3: fnu=Black Body @ T=10000K
                  4: for MIPS (leff with nu fnu=ctt and flux with BB @ 10000K
    FILTER_FILE
                output filter filename, will be saved in $LEPHAREWORK/filt/
    """

    def update_help(self):
        """Add specific help information"""
        doc = "Build the LePHARE internal representation of the set of filters to be used\n"
        with suppress(Exception):
            self.parser.usage = doc
        self.__doc__ = doc + "\n"  # + inspect.getdoc(Filter)

    def __init__(self, config_file=None, config_keymap=None, **kwargs):
        super().__init__(config_keys, config_file, config_keymap, **kwargs)

    def run(self, **kwargs):
        """Update keymap and verbosity based on call arguments.

        This is only when the code is called from python session.
        """

        super().run(**kwargs)

        keymap = self.keymap
        flt_rep = keymap["FILTER_REP"].split_string(os.path.join(os.environ["LEPHAREDIR"], "filt/"), 1)[0]
        flt_files = keymap["FILTER_LIST"].split_string("flt.pb", -99)
        ref_size = len(flt_files)
        # Transmission in energy or photons
        transtyp = (keymap["TRANS_TYPE"]).split_int("0", ref_size)
        # calibration depending on the instrument
        calibtyp = (keymap["FILTER_CALIB"]).split_int("0", ref_size)
        # Output file
        output_name = keymap["FILTER_FILE"].split_string("filters", 1)[0]
        filtfile = os.path.join(os.environ["LEPHAREWORK"], "filt", output_name + ".dat")
        filtdoc = os.path.join(os.environ["LEPHAREWORK"], "filt", output_name + ".doc")

        if self.verbose:
            print("#######################################")
            print("# Build the filter file with the following options: ")
            print(f"# Config file: {self.config}")
            print(f"# FILTER_REP: {flt_rep}")
            print(f"# FILTER_LIST: {flt_files}")
            print(f"# TRANS_TYPE: {transtyp}")
            print(f"# FILTER_CALIB: {calibtyp}")
            print(f"# FILTER_FILE: {filtfile}")
            print(f"# FILTER_FILE.doc: {filtdoc}")
            print("#######################################")

        vec_flt = []
        for k, (f, t, c) in enumerate(zip(flt_files, transtyp, calibtyp)):
            flt_file = f if os.path.isabs(f) else os.path.join(flt_rep, f)
            one_filt = flt(k, flt_file, t, c)
            vec_flt.append(one_filt)

        write_output_filter(filtfile, filtdoc, vec_flt)
        return vec_flt


def main():  # pragma no cover
    runner = Filter()
    _ = runner.run()
    runner.end()


if __name__ == "__main__":  # pragma no cover
    main()
